addressable.gemspec

```diff
@@ -27,9 +27,11 @@ Gem::Specification.new do |s|
 
   if s.respond_to? :add_runtime_dependency then
     s.add_runtime_dependency(%q<public_suffix>.freeze, [">= 2.0.2", "< 6.0"])
+    s.add_runtime_dependency(%q<ffi>.freeze)
     s.add_development_dependency(%q<bundler>.freeze, [">= 1.0", "< 3.0"])
   else
     s.add_dependency(%q<public_suffix>.freeze, [">= 2.0.2", "< 6.0"])
+    s.add_dependency(%q<ffi>.freeze)
```

Comment:
This is optional, we could also decide to make `ffi` optional like the `idn-ruby` gem. Obviously open to discussion, my opinion is that:
- leaving `ffi` as an optional dependency could create surprises. For example I add another gem to my project which installs `ffi` and suddenly `addressable` behavior "changes" because it uses `libidn2`. Or the other way around, which could cause an unexpected performance drop.
- `ffi` being a quite common and well maintained ruby gem it's not a big risk
- also as opposed to `idn-ruby`, `ffi` provides `libidn2` which is the up-to-date implementation, hopefully matching pure ruby version perfeclty once #491 is addressed, but likely faster.

So IMO it's ok to add it as a runtime dependency so people benefit from it if available.

Reply:
There currently is a [rake task](https://github.com/sporkmonger/addressable/blob/e91b64e1ab36dccacd8e339a84e89684ed5bbb3e/tasks/gem.rake#L5-L50) to that the gemspec is generated from that needs to be update in order to preserve these changes. (I'm not a big fan of this rake task, any new contributor doesn't know about it and I even forget it myself sometimes, and it specifies things you don't need to specify in the gemspec, and the [file listing is strange](https://github.com/sporkmonger/addressable/pull/497) (should change it to use git)).

I think you're right we should depend on `ffi` to make it as simple as possible to use `libidn2`... with that said, how are we going to roll out this change? In a minor version? In a major version? Should it be opt-in or opt-out to use `libidn2`?

Reply:
Aha indeed I missed this one ^^ this is the older way of doing gems so probably just legacy.
Also not a big fan of this way of doing, but let's not mix such independant change in this branch for the moment.

I addressed this in 0a6f09194a66f07e0b095ce87c844848547d497a by:
- Adding the `ffi` dependency in the rake task
- Updating the file regexp to remove the useless folders and add back the missing files
- Run the task and compared the output with main to make sure the only difference is `native2.rb` and `ffi` dependency. (aside from date and rubygems version ^^)
So now running `gem:release` should not cause any unwanted change.

About the behavior I believe it should be opt-out like it's currently the case with libidn, if it's available we just use it because it's supposed to be the "best" implementation. Fallback to `libidn` if available, and `pure` otherwise.
And if somebody wants to force one mode explicitely, I would say they can use something like `Addressable::IDNA = Addressable::IDNA::Pure` as suggested in my other comment (I'll add this to the readme if we do that)

About the version bump that is a good question, even though it's not breaking the API, it would still change the internal behavior a bit (especially IDNA2008+UTS46), some class names, etc. So I would say it could deserve a major version bump. But in that case, we should probably bundle #491 in the same major version bump, this way we ship one major update which brings IDNA2008+UTS46 to BOTH implementations at the same time, with the ability to still use the older IDNA2003 using `libidn` for people who needs. WDYT?

Reply:
Everything above sounds good to me :) 👍

Reply:
One alternative idea is adding this (`libidn2` support) in a minor version as opt-in (and make it opt-out (with #491) in a future major version), the pitch being it would be possible for users to prepare for the major version by upgrading to the minor version and switching backend back and forth. Does it makes sense or would it be too much work?

Reply:
Do have any other feedback on this part?

Reply:
I think what you did was great :)